### ISSUE TITLE
when there is a team/key problem, TSA should tell worker why the registration process exited

### ISSUE THREAD

**Issue by jamieklassen** [concourse/concourse#6196]
## Summary

Before starting, clone this repo and run
```
mkdir -p keys /tmp/work
go run ./cmd/concourse generate-key -t ssh -f keys/tsa
go run ./cmd/concourse generate-key -t ssh -f keys/worker
```
Then use the following `docker-compose.yml`:
```
version: '3'

services:
  db:
    image: postgres:${POSTGRES_TAG:-latest}
    environment:
      POSTGRES_DB: concourse
      POSTGRES_USER: dev
      POSTGRES_PASSWORD: dev
  web:
    image: concourse/concourse:6.6.0
    volumes:
    - ./keys:/keys
    command: web
    depends_on: [db]
    ports:
    - 2222:2222
    environment:
      CONCOURSE_POSTGRES_HOST: db
      CONCOURSE_POSTGRES_USER: dev
      CONCOURSE_POSTGRES_PASSWORD: dev
      CONCOURSE_POSTGRES_DATABASE: concourse
      CONCOURSE_ADD_LOCAL_USER: test:test,guest:guest
      CONCOURSE_MAIN_TEAM_LOCAL_USER: test
      CONCOURSE_TSA_TEAM_AUTHORIZED_KEYS: main:/keys/worker.pub
      CONCOURSE_TSA_HOST_KEY: /keys/tsa
```
and attempt
```
go run ./cmd/concourse worker --tsa-host localhost:2222 --tsa-worker-private-key keys/worker --work-dir /tmp/work --tsa-public-key keys/tsa.pub
```
and the process will exit 1, outputting
```
{"timestamp":"<timestamp>","level":"info","source":"worker","message":"worker.beacon-runner.exiting","data":{"reason":"registration process exited via SSH gateway","session":"<session>"}}
```
but why did the SSH gateway exit? If we inspect the tsa logs with `docker-compose logs web | grep tsa` we'll find one that stands out:
```
web_1  | {"timestamp":"<timestamp>","level":"error","source":"tsa","message":"tsa.connection.channel.exited-with-error","data":{"error":"key is authorized for team main, but worker is global","remote":"<remote address>","session":"<session>"}}
```
This is a pretty important error message. If I have that private key, I think the TSA should trust me to know what team it is for - others may disagree. But it would be very helpful for folks operating an external worker to receive this error -- otherwise the situation is pretty opaque and they need to call up whoever has access to the TSA logs!

Instead, I'd hope that when the worker exits, it would first spit out a log like
```
{"timestamp":"<timestamp>","level":"info","source":"worker","message":"worker.beacon-runner.exiting","data":{"reason":"registration process exited via SSH gateway: key is authorized for team main, but worker is global","session":"<session>"}}
```

## Context

Perhaps we could add a 'message' field to

https://github.com/concourse/concourse/blob/0e8560d9b2ca285870a752a43fa0954642ac7297/tsa/tsacmd/messages.go#L7-L9

? On the worker side we see there is a type assertion where the value is ignored:

https://github.com/concourse/concourse/blob/0e8560d9b2ca285870a752a43fa0954642ac7297/worker/beacon_runner.go#L60

whereas the `ExitError` type in the `golang.org/x/crypto/ssh` package does seem to support a fairly detailed message, including a 'reason':

https://github.com/golang/crypto/blob/9e8e0b390897c84cad53ebe9ed2d1d331a5394d9/ssh/session.go#L599-L647

It's not totally clear to me how our `exitStatusRequest` goes over the wire to become an `ExitError` on the worker side. It seems like there is some fairly low-level marshalling done by the ssh library.

**Comment by taylorsilva** [comment]
# Issue Bankruptcy

This issue is being closed as part of 2025's **Issue Bankruptcy** declaration. [See here for details](https://blog.concourse-ci.org/posts/2025-01-06-concourse-2024-review/)

If you think this issue still has merit then feel free to open a new issue and reference this one.
